hamilton/plugins/dlt_extensions.py

```diff
@@ -0,0 +1,74 @@
+import dataclasses
+from typing import Any, Collection, Dict, Iterable, Literal, Optional, Type
+
+import dlt
+import pandas as pd
+
+# importing TDestinationReferenceArg fails if Destination isn't imported
+from dlt.common.destination import Destination, TDestinationReferenceArg  # noqa: F401
+
+from hamilton import registry
+from hamilton.io.data_adapters import DataSaver
+
+try:
+    import pyarrow as pa
+except ModuleNotFoundError:
+    SAVER_TYPES = [Iterable, pd.DataFrame]
+else:
+    SAVER_TYPES = [Iterable, pd.DataFrame, pa.Table, pa.RecordBatch]
+
+
+@dataclasses.dataclass
+class DltSaver(DataSaver):
+    """Materialize results using a dlt pipeline with the specified destination.
+
+    In reference to an Extract, Transform, Load (ETL) pipeline, here, the Hamilton
+    dataflow is responsible for Transform, and `DltDestination` for Load.
+    """
+
+    pipeline: dlt.Pipeline
+    rename: Optional[dict] = None
+    # kwargs for pipeline.run()
+    write_disposition: Optional[Literal["skip", "append", "replace", "merge"]] = None
+
+    @classmethod
+    def applicable_types(cls) -> Collection[Type]:
+        return SAVER_TYPES
+
+    def save_data(self, data) -> Dict[str, Any]:
+        """
+        ref:
+            tables: https://dlthub.com/docs/dlt-ecosystem/verified-sources/arrow-pandas
+            standalone resources: https://dlthub.com/docs/general-usage/resource#load-resources
+            dynamic resources: https://dlthub.com/docs/general-usage/source#create-resources-dynamically
+        """
+        if isinstance(data, dict) is False:  # only 1 node, no `combine=base.DictResult()`
```

Comment:
So this `data` type maps to the `applicable_types`. As in, this is the type that we can save.

The `combiner` will create a node that will have an output type of the corresponding result builder. E.G. the `DataFrameResult()`. So, this check shoudln't be necessary if `dict` is the only one. Also it means that none of the other ones should work.

That said, the ergonomics feel off... What are the use-cases behind the other types?